core/src/Streamly/Internal/FileSystem/Posix/File.hsc

```diff
@@ -0,0 +1,318 @@
+module Streamly.Internal.FileSystem.Posix.File
+    (
+#if !defined(mingw32_HOST_OS) && !defined(__MINGW32__)
+
+    -- * File open flags
+      OpenFlags (..)
+    , defaultOpenFlags
+
+    -- * File status flags
+    , setAppend
+    , setNonBlock
+    , setSync
+
+    -- * File creation flags
+    , setCloExec
+    , setDirectory
+    , setExcl
+    , setNoCtty
+    , setNoFollow
+    -- setTmpFile
+    , setTrunc
+
+    -- * File create mode
+    , defaultCreateMode
+
+    -- ** User Permissions
+    , setUr
+    , setUw
+    , setUx
+
+    , clrUr
+    , clrUw
+    , clrUx
+
+    -- ** Group Permissions
+    , setGr
+    , setGw
+    , setGx
+
+    , clrGr
+    , clrGw
+    , clrGx
+
+    -- ** Other Permissions
+    , setOr
+    , setOw
+    , setOx
+
+    , clrOr
+    , clrOw
+    , clrOx
+
+    -- ** Status bits
+    , setSuid
+    , setSgid
+    , setSticky
+
+    , clrSuid
+    , clrSgid
+    , clrSticky
+
+    -- * Fd based Low Level
+    , openAt
+    , close
+
+    -- * Handle based
+    , openFile
+    , withFile
+    -- , openBinaryFile
+    -- , withBinaryFile
+
+    -- Re-exported
+    , Fd
+#endif
+    ) where
+
+#if !defined(mingw32_HOST_OS) && !defined(__MINGW32__)
+
+-------------------------------------------------------------------------------
+-- Imports
+-------------------------------------------------------------------------------
+
+import Data.Bits ((.|.), (.&.), complement)
+import Foreign.C.Error (throwErrnoIfMinus1_)
+import Foreign.C.String (CString)
+import Foreign.C.Types (CInt(..))
+import GHC.IO.Handle.FD (fdToHandle)
+import Streamly.Internal.FileSystem.Posix.Errno (throwErrnoPathIfMinus1Retry)
+import Streamly.Internal.FileSystem.PosixPath (PosixPath)
+import System.IO (IOMode(..), Handle)
+import System.Posix.Types (Fd(..), CMode(..))
+
+import qualified Streamly.Internal.FileSystem.File.Common as File
+import qualified Streamly.Internal.FileSystem.PosixPath as Path
+
+-- We want to remain close to the Posix C API. A function based API to set and
+-- clear the modes is simple, type safe and directly mirrors the C API. It does
+-- not require explicit mapping from Haskell ADT to C types, we can dirctly
+-- manipulate the C type.
+
+#include <fcntl.h>
+
+-------------------------------------------------------------------------------
+-- Create mode
+-------------------------------------------------------------------------------
+
+-- | Open flags, see posix open system call man page.
+newtype FileMode = FileMode CMode
+
+##define MK_MODE_API(name1,name2,x) \
+{-# INLINE name1 #-}; \
+name1 :: FileMode -> FileMode; \
+name1 (FileMode mode) = FileMode (x .|. mode); \
+{-# INLINE name2 #-}; \
+name2 :: FileMode -> FileMode; \
+name2 (FileMode mode) = FileMode (x .|. mode)
+
+{-
```

Comment:
Can we even comment CPP definitions?

Reply:
No, we are using the definitions directly from the header files rather than defining ourselves. I kept this code commented, just in case there is any issue in using the headers directly.

Reply:
But if it did not error out because you've not used `##` it likely works as expected.

Reply:
This code is actually commented, it does not work in any way.